sway-lib-std/src/contract_id.sw

```diff
@@ -65,6 +65,120 @@ impl From<b256> for ContractId {
     }
 }
 
+impl Hash for ContractId {
+    fn hash(self, ref mut state: Hasher) {
+        let ContractId { value } = self;
+        value.hash(state);
+    }
+}
+
+/// An AssetId is used for interacting with an asset on the network. 
+///
+/// # Additional Information
+///
+/// It is calculated by taking the sha256 hash of the originating ContractId and a SubId.
+/// ie. sha256((contract_id, sub_id)).
+///
+/// An exception is the Base Asset, which is just the ZERO_B256 AssetId.
+///
+/// The SubId is used to differentiate between different assets that are created by the same contract.
+pub struct AssetId {
+    value: b256,
+}
+
+impl core::ops::Eq for AssetId {
+    fn eq(self, other: Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl AssetId {
+    /// Creates a new AssetId from a ContractId and SubId.
+    ///
+    /// # Arguments
+    ///
+    /// * `contract_id`: [ContractId] - The ContractId of the contract that created the asset.
+    /// * `sub_id`: [SubId] - The SubId of the asset.
+    ///
+    /// # Returns
+    ///
+    /// * [AssetId] - The AssetId of the asset. Computed by hashing the ContractId and SubId.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// use std::{callframes::contract_id, constants::ZERO_B256};
+    ///
+    /// fn foo() {
+    ///     let contract_id = contract_id();
+    ///     let sub_id = ZERO_B256;
+    ///
+    ///     let asset_id = AssetId::new(contract_id, sub_id);        
+    /// }
+    /// ```
+    pub fn new(contract_id: ContractId, sub_id: SubId) -> Self {
+        let value = sha256((contract_id, sub_id));
+        Self { value }
+    }
+
+    /// Creates a new AssetId from a ContractId and the zero SubId.
+    ///
+    /// # Arguments
+    ///
+    /// * `contract_id`: [ContractId] - The ContractId of the contract that created the asset.
+    ///
+    /// # Returns
+    ///
+    /// * [AssetId] - The AssetId of the asset. Computed by hashing the ContractId and the zero SubId.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// use std::{callframes::contract_id, constants::ZERO_B256};
+    ///
+    /// fn foo() {
+    ///     let contract_id = contract_id();
+    ///     let sub_id = ZERO_B256;
+    ///
+    ///     let asset_id = AssetId::standard(contract_id);
+    ///
+    ///     assert(asset_id == AssetId::new(contract_id, sub_id));
+    /// }
+    /// ```
+    pub fn standard(contract_id: ContractId) -> Self {
+        let value = sha256((contract_id, 0x0000000000000000000000000000000000000000000000000000000000000000));
```

Comment:
Can this be the ZERO_B256 constant?

Reply:
no, that file needs to import this file

Reply:
+1 to @bitzoic's comment

Reply:
Its an internal implementation detail, plenty of other places do this to avoid circular imports